client/v2/tx/factory.go

```diff
@@ -0,0 +1,588 @@
+type Factory struct {
+	keybase           keyring.Keyring
+	txBuilderProvider TxBuilderProvider
+	codec             TxCodecProvider
+	accountRetriever  AccountRetriever
+	signingConfig     TxSigningConfig
+	TxParameters
+}
+
+func NewFactoryCLI(clientCtx txContext, flagSet *pflag.FlagSet) (Factory, error) {
+	var accNum, accSeq uint64
+	if clientCtx.Offline {
+		if flagSet.Changed(flags.FlagAccountNumber) && flagSet.Changed(flags.FlagSequence) {
+			accNum = clientCtx.Viper.GetUint64(flags.FlagAccountNumber)
+			accSeq = clientCtx.Viper.GetUint64(flags.FlagSequence)
+		} else {
+			return Factory{}, fmt.Errorf("account-number and sequence must be set in offline mode")
+		}
+	}
+
+	if clientCtx.Offline && clientCtx.GenerateOnly {
+		if clientCtx.ChainID != "" {
+			return Factory{}, errors.New("chain ID cannot be used when offline and generate-only flags are set")
+		}
+	} else if clientCtx.ChainID == "" {
+		return Factory{}, errors.New("chain ID required but not specified")
+	}
+
+	signMode := flags.ParseSignMode(clientCtx.SignModeStr)
+	memo := clientCtx.Viper.GetString(flags.FlagNote)
+	timeoutHeight := clientCtx.Viper.GetUint64(flags.FlagTimeoutHeight)
+	unordered := clientCtx.Viper.GetBool(flags.FlagUnordered)
+
+	gasAdj := clientCtx.Viper.GetFloat64(flags.FlagGasAdjustment)
+	gasStr := clientCtx.Viper.GetString(flags.FlagGas)
+	gasSetting, _ := flags.ParseGasSetting(gasStr)
+	gasPricesStr := clientCtx.Viper.GetString(flags.FlagGasPrices)
+
+	feesStr := clientCtx.Viper.GetString(flags.FlagFees)
+
+	f := Factory{
+		accountRetriever: clientCtx.AccountRetriever,
+		keybase:          clientCtx.Keyring,
+		TxParameters: TxParameters{
+			timeoutHeight: timeoutHeight,
+			memo:          memo,
+			chainID:       clientCtx.ChainID,
+			signMode:      signMode,
+			AccountConfig: AccountConfig{
+				accountNumber: accNum,
+				sequence:      accSeq,
+				fromName:      clientCtx.FromName,
+				fromAddress:   clientCtx.FromAddress,
+			},
+			GasConfig: GasConfig{
+				gas:           gasSetting.Gas,
+				gasAdjustment: gasAdj,
+			},
+			FeeConfig: FeeConfig{
+				feeGranter: clientCtx.FeeGranter,
+				feePayer:   clientCtx.FeePayer,
+			},
+			ExecutionOptions: ExecutionOptions{
+				unordered:          unordered,
+				offline:            clientCtx.Offline,
+				generateOnly:       clientCtx.GenerateOnly,
+				simulateAndExecute: gasSetting.Simulate,
+				preprocessTxHook:   clientCtx.PreprocessTxHook,
+			},
+		},
+	}
+
+	f = f.WithFees(feesStr).WithGasPrices(gasPricesStr)
+	return f, nil
+}
+
+// Prepare ensures the account defined by ctx.GetFromAddress() exists and
+// if the account number and/or the account sequence number are zero (not set),
+// they will be queried for and set on the provided Factory.
+// A new Factory with the updated fields will be returned.
+// Note: When in offline mode, the Prepare does nothing and returns the original factory.
+func (f Factory) Prepare(clientCtx txContext) (Factory, error) {
```

Comment:
Idea: what if the `Prepare` func does all the online logics besides setting accNum, accSeq all the gas related online queries ?. IMO having a clear separation between online and offline functions makes is easier to maintain.